### ISSUE TITLE
Remove YouTube refresh hack (after menu opened) and fix it

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#393]
It's an annoying behavior on youtube.

The problem is that YouTube doesn't gain focus. See #385 (and it's bug) for more details.

**Comment by bbinto** [comment]
Let's wait until @tif proposes her UI improvements

**Comment by mcomella** [comment]
Waiting for the new overlay to see if it's still a bug.

**Comment by liuche** [comment]
Quick update on using the overlay - this doesn't fix the problem, even though the views that originally had focus are gone, the focus doesn't get "fixed". Still some things I have in mind to try while I finish #444 , but just swapping the overlay doesn't fix this.

**Comment by Sdaswani** [comment]
That is no bueno. 😡

**Comment by liuche** [comment]
Current state with the new overlay: hack is still in place, nothing has changed or improved

On youtube.com/tv, close the overlay. Youtube force-refreshes (so that we can get remote input and move around, pick videos, etc).

**Comment by Sdaswani** [comment]
@bbinto I think we can take this out of 2.0 since the new overlay doesn't change the refresh situation with youtube - finding a fix is non-trivial.

**Comment by bbinto** [comment]
:( ok, makes sense

**Comment by liuche** [comment]
Might be fixed by #654

**Comment by mcomella** [comment]
I can reproduce this in a standalone app on the emulator:
- Start WebView, sleep 2s
- Focus another view, sleep 2s
- Focus webView

And the WebView controls cannot be accessed anymore. Here's my standalone app code to reproduce: https://github.com/mcomella/firefox-tv-issue393/blob/0cac8fdb97a3bcc8f3c30379f31f250d2f6e5a4d/app/src/main/java/xyz/mcomella/webviewevaljs/MainActivity.kt#L22

I think it'll be much faster to fix in this environment than on the TV.

**Comment by mcomella** [comment]
Clicking the left sidebar on the page (with a touch event) appears to fix the focus issue.

**Comment by mcomella** [comment]
In JS, calling `DOMElement.focus()` (e.g. on the element that the user has currently focused) also fixes this issue.

**Comment by mcomella** [comment]
If I load google.com instead, I can still navigate with the keyboard after the WebView gets focus again: it seems like **this is actually youtube.com/tv 's fault**.

Caveat: when I repro on google.com, when the WebView regains focus, the element I currently had focused is no longer focused and nothing appears to be focused. Maybe the problem is that Youtube.com/tv does not expect to lose focus.

**Comment by mcomella** [comment]
When the WebView is unfocused and refocused, the active DOMElement is `<body>` rather than whatever the user had focused before: **it's the WebView's fault that it doesn't refocus the existing element and it's youtubeTV's fault for not handling that case.**

**Comment by mcomella** [comment]
My solution was basically, in WebViewProvider.initWebview:
```kotlin
    setOnFocusChangeListener { _, hasFocus ->
        // TODO: docs. WHy never called?
        // TODO: It's attached. Maybe it's because it's IWebView?
        if (!hasFocus) {
            // TODO: only if cursor active site.
            Log.d("lol", "unfocus")
            evalJS("var _lol = document.activeElement;")
        } else {
            Log.d("lol", "focused")
            evalJS("_lol.focus();")
        }
    }
```

But the focus change listener is never called despite:
1) The getter returning the right OnFocusChangedListener object
2) It losing focus, according to the layout inspector.

My going theory is that Android is manipulating IWebView, which it doesn't consider a View (since it doesn't extend View), so it doesn't call onFocusChanged. Though `MainActivity.onCreateView` does return `View?`... so maybe that's not true.

fwiw, this solution does work in a standalone app on Android's WebView.

@pocmo Have you ever encountered issues with setting focus listeners on IWebView?

**Comment by Sdaswani** [comment]
@mcomella I think this is a important fix so I would continue on if you think you are close to a solution.

**Comment by pocmo** [comment]
> @pocmo Have you ever encountered issues with setting focus listeners on the WebView underlying IWebView?

No idea, sorry.

> fwiw, this solution does work in a standalone app on Android's WebView.

Just to be sure: Does your standalone app use the same base activity and/or fragment? At least AppCompatActivity does some magic too in order to replace some view classes with its own implementation, I think.

**Comment by mcomella** [comment]
> fwiw, this solution does work in a standalone app on Android's WebView.

I tried with AmazonWebView in a raw app and the onFocusChangeListener is never called – it appears the problem is with AmazonWebView. I also tried overriding the `AmazonWebView.onFocusChanged` and `onWindowFocusChanged` and those aren't called either.

**Comment by mcomella** [comment]
I was able to fix the problem in the default use case with the following code:
```kotlin
    viewTreeObserver.addOnGlobalFocusChangeListener { oldFocus: View?, newFocus: View? ->
        fun View?.isWebView() = if (this == null) false else webView.anyChild { child -> this == child }

        if (oldFocus.isWebView()) {
            Log.d("lol", "unfocus")
            evalJS("var _lol = document.activeElement;")
            evalJS("console.log('lol: ' + _lol);")
        } else if (newFocus.isWebView()) {
            Log.d("lol", "focused")
            h.post { // Needs to run later on UI thread
                evalJS("_lol.focus();")
                evalJS("console.log('lol: ' + _lol);")
            }
        }
    }
```

But it doesn't work when we refresh the page or go back because we didn't have a chance to retrieve the focused element (i.e. when losing focus).

To work around this, in addition to the above code, we could:
- Try to find another entry point for capturing the focused View (e.g. if navigationEvent occurs, add `onLoad` JS listener that stores the view, assuming the correct view is focused; or try when the overlay is hidden - this is getting pretty messy)
- Make a hack for YouTube to fix the focused state based on the CSS, which indicates which view is focused. YouTube is the only website we've found that breaks when not focused (and it's the only one we give keyboard nav to) so this could be a fine approach

**Comment by mcomella** [comment]
Okay, I got a solution that seems reasonable – I just need to clean it up.

**Comment by Sdaswani** [comment]
Awesome!!!!

**Comment by mcomella** [comment]
> it's the WebView's fault that it doesn't refocus the existing element

I filed https://bugs.chromium.org/p/chromium/issues/detail?id=826577

**Comment by liuche** [comment]
@mcomella looks like your Chromium issue was triaged to be a Chrome issue - maybe you weren't clear that you were using the WebView rather than Chrome itself.

**Comment by mcomella** [comment]
This was fixed with #682.

**Comment by mcomella** [comment]
QA: Before this fix and without our old hack to reload youtube.com/tv, opening and closing the overlay would cause the dpad to be unresponsive on youtube.com/tv, e.g. STR:
1) Open youtube.com/tv
2) Click menu button on controller to open overlay
3) Click menu button on controller to close overlay
4) Click right on dpad

Expected: move right
Actual: move nowhere. dpad isn't working.

It'd be great if you can verify that you can't reproduce the behavior where the dpad stops working on youtube.com/tv in any way. Also, quickly check that you don't see weird behavior on other pages that use the Cursor: to fix this issue, we added javascript code to focus DOM elements on *all* pages so I want to verify it has no unexpected effects.
